### PR TITLE
abort_source: subscription: allow calling on_abort explicitly

### PR DESCRIPTION
**Pull request by bhalevy**
Today, when calling `subscribe` after abort has alredy been requested on an `abort_source` we return a default- initialized subscription that signifies that subscription failed.

With this change, `subscribe` would return an unlinked subscription, that still holds the required callback. `bool(subscription)` stil elaluates to `false` in the same way, since the subscription is unlinked, but, it supports `on_abort`, that calls the subscribed
callback, at a later time, the same way as if
abort is requested at a later time.

Thise allows the users of this interface to
implement a unified abort path that can either be called by the abort_source, if subscribe happened before
abort was requested, or called by the user `on_abort` call, if abort weas never requested, or if it was requested before `subscribe` was called.

### CODE REVIEW

File: include/seastar/core/abort_source.hh [hunk 1]
```diff
@@ -140,9 +149,6 @@ public:
                   std::is_nothrow_invocable_r_v<void, Func>)
     [[nodiscard]]
     optimized_optional<subscription> subscribe(Func&& f) {
-        if (abort_requested()) {
-            return { };
-        }
         if constexpr (std::is_invocable_v<Func, std::exception_ptr>) {
```
Comment (avikivity) [inline include/seastar/core/abort_source.hh:152]: Isn't the documentation no longer correct?

Comment (bhalevy) [reply include/seastar/core/abort_source.hh:152]: Good catch. I'll address that in the next version.

**Review by xemul — APPROVED** [review]

File: include/seastar/core/abort_source.hh (hunk 1 above)
Comment (bhalevy) [reply include/seastar/core/abort_source.hh:167]: fixed in v2

**Comment by bhalevy** [comment]
v2: improve documentation

**Comment by bhalevy** [comment]
@xemul is there no CI?

**Comment by xemul** [comment]
> @xemul is there no CI?

@bhalevy , I don't follow. Looks like PR's CI was started with delay, or do you mean something else?

**Comment by bhalevy** [comment]
> > @xemul is there no CI?
> 
> @bhalevy , I don't follow. Looks like PR's CI was started with delay, or do you mean something else?

Ah no.  I looked after force pushing a new version, so it probably erased the previous status and showed the new CI invocation with a delay of few minutes, as you mentioned.
I was concerned that CI didn't run even for v1...
The bottom line is that it's running now! :)

**Review by Deexie — APPROVED** [review]

**Comment by bhalevy** [comment]
@avikivity / @scylladb/seastar-maint can you please merge this PR?

**Comment by tchaikov** [comment]
@avikivity / @scylladb/seastar-maint  ping?

**Pull request closed without merging**